coffee_core/src/coffee.rs

```diff
@@ -246,13 +247,37 @@ impl PluginManager for CoffeeManager {
         plugin: &str,
         verbose: bool,
         try_dynamic: bool,
+        branch: Option<String>,
     ) -> Result<(), CoffeeError> {
         log::debug!("installing plugin: {plugin}");
         // keep track if the plugin is successfully installed
         for repo in self.repos.values() {
             if let Some(mut plugin) = repo.get_plugin_by_name(plugin) {
                 log::trace!("{:?}", plugin);
 
+                // if the branch is specified, we need to save current git_head
+                // and checkout the branch
+                let current_git_head = repo.git_head();
+                let current_plugin_commit = plugin.commit.clone();
+                if let Some(branch) = branch.clone() {
+                    log::debug!("checking out branch: {branch}");
+                    if current_git_head.is_none() {
+                        return Err(error!("unable to get the current git head"));
+                    }
+                    let updated_status =
+                        git_upgrade_with_branch(repo.url().path_string.as_str(), &branch, verbose)
+                            .await?;
+                    log::debug!("updated status: {:?}", updated_status);
+
+                    let repository = git2::Repository::open(repo.url().path_string.as_str())
+                        .map_err(|err| error!("{}", err.message()))?;
+
+                    // get commit id after the checkout
+                    let commit = commit_id!(repository);
+                    log::debug!("commit id after the checkout: {commit}");
+                    plugin.commit = Some(commit);
+                }
```

Comment:
IMHO to much code for just `sh!("git checkout {branch}")`?

Reply:
I can see some potential issues with simply using `git checkout {branch}`:
- If the branch is new (not fetched yet), we need to `git fetch origin` first. 
	- To make sure we are returning to the expected state after switching back, we need the repository's Git HEAD, not just the current branch. This is because the branch might have been updated remotely (when we ran `git fetch origin`), and we want to go back to the exact commit the user was on.
	-  We need to verify that the repository's Git HEAD information exists (introduced in a previous PR as an Option to be backward compatible) 
- The plugin's `commit` field is updated during installation and then reverted afterward. Otherwise, we have false information.

Do you think that any of these steps is redundant?

Reply:
`git_upgrade_with_branch()` does `git fetch origin`